File: app/src/main/java/chenmc/sms/utils/ToastUtil.java
package chenmc.sms.utils;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.StringRes;
import android.widget.Toast;

import chenmc.sms.ui.app.App;

/**
 * @author 明 明
 * Created on 2017-4-30.
 */

public class ToastUtil {
    
    private static final String KEY_TEXT = "text";
    private static final String KEY_DURATION = "duration";
    
    private static Handler mHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            Bundle data = msg.getData();
            
            Toast.makeText(App.Companion.getContext(), data.getString(KEY_TEXT),
                data.getInt(KEY_DURATION) ==
                Toast.LENGTH_SHORT ? Toast.LENGTH_SHORT : Toast.LENGTH_LONG
            ).show();
        }
    };
    
    public static void showToast(String text, int duration) {
        Message message = mHandler.obtainMessage();
        Bundle bundle = new Bundle();
        bundle.putString(KEY_TEXT, text);
        bundle.putInt(KEY_DURATION, duration);
        message.setData(bundle);
        message.sendToTarget();
    }
    
    public static void showToast(@StringRes int textRes, int duration) {
        showToast(App.Companion.getContext().getString(textRes), duration);
    }
    
}

File: app/src/main/java/chenmc/sms/util/ToastUtil.kt
package chenmc.sms.util

import android.widget.Toast
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import chenmc.sms.ui.app.App

object ToastUtil {

    private val mToast = Toast.makeText(App.context, "", Toast.LENGTH_SHORT)

    @IntDef(value = [Toast.LENGTH_SHORT, Toast.LENGTH_LONG])
    @Retention(AnnotationRetention.SOURCE)
    annotation class Duration

    fun showSingletonToast(text: String, @Duration duration: Int) {
        mToast.setText(text)
        mToast.duration = duration
        mToast.show()
    }

    fun showSingletonToast(@StringRes textRes: Int, @Duration duration: Int) {
        mToast.setText(textRes)
        mToast.duration = duration
        mToast.show()
    }

    fun showSingletonShortToast(text: String) {
        showSingletonToast(text, Toast.LENGTH_SHORT)
    }

    fun showSingletonShortToast(@StringRes textRes: Int) {
        showSingletonToast(textRes, Toast.LENGTH_SHORT)
    }

    fun showSingletonLongToast(text: String) {
        showSingletonToast(text, Toast.LENGTH_LONG)
    }

    fun showSingletonLongToast(@StringRes textRes: Int) {
        showSingletonToast(textRes, Toast.LENGTH_LONG)
    }

    fun showToast(text: String, @Duration duration: Int) {
        Toast.makeText(App.context, text, duration).show()
    }

    fun showToast(@StringRes textRes: Int, @Duration duration: Int) {
        Toast.makeText(App.context, textRes, duration).show()
    }

    fun showShortToast(text: String) {
        showToast(text, Toast.LENGTH_SHORT)
    }

    fun showShortToast(@StringRes textRes: Int) {
        showToast(textRes, Toast.LENGTH_SHORT)
    }

    fun showLongToast(text: String) {
        showToast(text, Toast.LENGTH_LONG)
    }

    fun showLongToast(@StringRes textRes: Int) {
        showToast(textRes, Toast.LENGTH_LONG)
    }
}